### ISSUE TITLE
AssertionProp incorrectly removes cast from uint

### ISSUE THREAD

**Issue by kalimag** [dotnet/coreclr#25134]
The following code prints a negative number instead of the expected `OverflowException` under certain conditions:

    class Program
    {
        static uint _value = int.MaxValue + 1U;

        static void Main()
        {
            try
            {
                var result = CastToIntChecked(_value);
                Console.WriteLine(result);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex);
            }
        }

        static int CastToIntChecked(uint value) => checked((int)value);
    }

This only seems to happen when `CastToIntChecked` is inlined. Preventing inlining by running a debug build, applying `MethodImplOptions.NoInlining` or bloating the method body results in an overflow exception.

Other changes that cause overflow checking to be performed as expected include:

- calling `CastToIntChecked(int.MaxValue + 1U)`

- calling `CastToIntChecked(_value + 1)`

- calling `CastToIntChecked(uint.Parse("2147483648"))`

- reading `_value` inside `CastToIntChecked` instead of passing it in as an argument

- performing other overflowing conversions like `ulong`->`long` or `ulong`->`int`

I haven't been able to extensively determine the affected runtimes, but as far as I can tell this occurs at least on Windows builds of .NET Core 2.2, 3.0-preview5, as well as .NET Desktop x64 but not .NET Desktop x86.

**Comment by RussKeldorph** [comment]
@dotnet/jit-contrib

**Comment by AndyAyersMS** [comment]
This looks like a bug in local assertion prop .. the subrange assertion should not allow the cast to be removed.
```
GenTreeNode creates assertion:
               [000029] -ACXG-------              *  ASG       int   
In BB01 New Local Subrange Assertion: V02 in [-2147483648..2147483647] index=#01, mask=0000000000000001

fgMorphTree BB01, stmt 1 (after)
               [000029] -ACXG+------              *  ASG       int   
               [000028] D----+-N----              +--*  LCL_VAR   int    V02 tmp2         
               [000009] --CXG+------              \--*  COMMA     int   
               [000008] H-CXG+------                 +--*  CALL help long   HELPER.CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
               [000004] -----+------ arg0 in rcx     |  +--*  CNS_INT   long   0x7ffc5495af18
               [000005] -----+------ arg1 in rdx     |  \--*  CNS_INT   int    1
               [000002] ----G+------                 \--*  CLS_VAR   int    Hnd=0x5495c2e0 Fseq[_value]

fgMorphTree BB01, stmt 2 (before)
               [000014] --C-G-------              *  CALL      void   System.Console.WriteLine
               [000027] ---X-----U-- arg0         \--*  CAST_ovfl int <- uint
               [000026] ------------                 \--*  LCL_VAR   int    V02 tmp2         
Initializing arg info for 14.CALL:
ArgTable for 14.CALL after fgInitArgInfo:
fgArgTabEntry[arg 0 27.CAST int, 1 reg: rcx, align=1]

Morphing args for 14.CALL:

Subrange prop for index #01 in BB01:
               [000027] ---X-----U--              *  CAST_ovfl int <- uint

fgMorphTree BB01, stmt 2 (after)
               [000014] --CXG+------              *  CALL      void   System.Console.WriteLine
               [000026] -----+------ arg0 in rcx  \--*  LCL_VAR   int    V02 tmp2         

```

**Comment by briansull** [comment]
I will investigate further

**Comment by AndyAyersMS** [comment]
I looked a bit more, we are not paying attention to the GTF_UNSIGNED on the cast.

It's not clear to me how to safely handle subrange prop in casts that also change signedness  -- seems like we would have to do some math and possibly swap high and low bounds -- so I was thinking we should just bail out if we see these cases.

**Comment by AndyAyersMS** [comment]
Also struck me as odd that `GenTree::CastFromType()` does not look at `GTF_UNSIGNED` either. Seems like this is just asking for trouble.

**Comment by mikedn** [comment]
> Also struck me as odd that GenTree::CastFromType() does not look at GTF_UNSIGNED either. Seems like this is just asking for trouble.

`CastFromType` should just be deleted. It's rarely used and all it does is to return the type of the cast operand. Not to mention that for unknown reasons it's on `GenTree` instead of `GenTreeCast`.

**Comment by briansull** [comment]
This issue can also be reproduced without inlining.

**Comment by briansull** [comment]
Resolved by #25260
